url294e6c45f3_Ecdosis_NMergeNew_tgz-pJ8-TestJ8/src/edu/luc/nmerge/fastme/edge.java

```diff
@@ -0,0 +1,257 @@
+	    		if (null != f)
+	    			f.updateSizes(direction.UP);
+	    		f = tail.parentEdge;
+	    		if ( null != f )
+	    			f.updateSizes( direction.DOWN );
+	    		bottomsize++;
+	    		break;
+	    }
+	}      
+	void assignBottomsize()
+	{
+		if (head.leaf())
+			bottomsize = 1;
+		else
+	    {
+			head.leftEdge.assignBottomsize();
+			head.rightEdge.assignBottomsize();
+			bottomsize = head.leftEdge.bottomsize
+				+ head.rightEdge.bottomsize;
+	    }
+	}
+	void assignTopsize( int numLeaves )
+	{
+		topsize = numLeaves - bottomsize;
+		head.leftEdge.assignTopsize(numLeaves);
+		head.rightEdge.assignTopsize(numLeaves);
+	}
+	/**
+	 * OLSint and OLSext use the average distance array to 
+	 * calculate weights instead of using the edge average weight 
+	 * fields
+	 */
+	void OLSext( double[][] A )
+	{
+		edge f, g;
+		if ( head.leaf())
+	    {
+			f = siblingEdge();
+			distance = 0.5*(A[head.index][tail.index] 
+				 + A[head.index][f.head.index]
+				 - A[f.head.index][tail.index]);
+	    }
+		else
+	    {
+			f = head.leftEdge;
+			g = head.rightEdge;
+			distance = 0.5*(A[head.index][f.head.index]
+				+ A[head.index][g.head.index]
+			    - A[f.head.index][g.head.index]);
+	    }
+	}
+	double wf( double lambda, double D_LR, double D_LU, double D_LD, 
+			   double D_RU, double D_RD, double D_DU )
+	{
+		double weight;
+		weight = 0.5*(lambda*(D_LU  + D_RD) + (1 -lambda)*(D_LD + D_RU)
+			- (D_LR + D_DU));  
+		return weight;
+	}
+	void OLSint( double[][] A)
+	{
+		double lambda;
+		edge left, right, sib;
+		left = head.leftEdge;
+		right = head.rightEdge;
+		sib = siblingEdge();
+		lambda = ((double) sib.bottomsize*left.bottomsize + 
+		    right.bottomsize*tail.parentEdge.topsize) /
+	    (bottomsize*topsize);
+		distance = wf(lambda,A[left.head.index][right.head.index],
+		    A[left.head.index][tail.index],
+		    A[left.head.index][sib.head.index],
+		    A[right.head.index][tail.index],
+		    A[right.head.index][sib.head.index],
+		    A[sib.head.index][tail.index]);
+	}
+	/**
+	 * works except when e is the one edge inserted to new vertex 
+	 * v by firstInsert
+	 */
+	void WFext( double[][] A) 
+	{
+		edge f, g;
+		if ( head.leaf() && tail.leaf() )
+			distance = A[head.index][head.index];
+		else if ( head.leaf() )
+		{
+			f = tail.parentEdge;
+			g = siblingEdge();
+			distance = 0.5*(A[head.index][g.head.index]
+			   + A[head.index][f.head.index]
+			   - A[g.head.index][f.head.index]);
+		}
+		else
+		{
+			f = head.leftEdge;
+			g = head.rightEdge;
+			distance = 0.5*(A[g.head.index][head.index]
+			     + A[f.head.index][head.index]
+			     - A[f.head.index][g.head.index]);
+		}
+	}
+	void WFint( double[][] A )
+	{
+		int up, down, left, right;
+		up = tail.index;
+		down = siblingEdge().head.index;
+		left = head.leftEdge.head.index;
+		right = head.rightEdge.head.index;
+		distance = 0.25*(A[up][left] + A[up][right] 
+		    + A[left][down] + A[right][down]) 
+		    - 0.5*(A[down][up] + A[left][right]);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 213:219, 249:255.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.